File: DS/BST/c/pr_01.c
// Bst

#include <stdio.h>
#include <stdlib.h>
struct Node
{
    struct Node *left;
    int data;
    struct Node *right;

} *root = NULL;

void Insert(int key)
{
    struct Node *t = root;
    struct Node *r, *p;

    if (root == NULL)
    {
        p = (struct Node *)malloc(sizeof(struct Node));
        p->data = key;
        p->left = p->right = NULL;
        root = p;
        return;
    }
    while (t != NULL)
    {
        r = t;
        if (key < t->data)
        {
            t = t->left;
        }
        else if (key > t->data)
        {
            t = t->right;
        }
        else
        {
            return;
        }
    }
    p = (struct Node *)malloc(sizeof(struct Node));
    p->data = key;
    p->left = p->right = NULL;
    if (key < r->data)
    {
        r->left = p;
    }
    else
    {
        r->right = p;
    }
}

struct Node *search(int key)
{
    struct Node *t = root;
    while (t != NULL)
    {
        if (key == t->data)
        {
            return t;
        }
        else if (key < t->data)
        {
            t = t->left;
        }
        else
        {
            t = t->right;
        }
    }
    return NULL;
}
void Inorder(struct Node *p)
{
    if (p)
    {
        Inorder(p->left);
        printf("%d ", p->data);
        Inorder(p->right);
    }
}

int main()
{
    Insert(10);
    Insert(5);
    Insert(20);
    Insert(8);
    Insert(30);
    Inorder(root);

    struct Node *temp = search(20);
    printf("\n");
    if (temp)
    {
        printf("Element is found at %d\n", temp->data);
    }
    else
    {
        printf("Not found");
    }
    printf("\n");
}

File: DS/BST/java/pr_01.java

// package BST.java;
import java.util.*;
import javax.sound.sampled.SourceDataLine;

public class pr_01 {
    static class Node {
        int data;
        Node left;
        Node right;

        Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    public static Node insert(Node root, int val) {
        if (root == null) {
            root = new Node(val);
            return root;
        }
        if (root.data > val) {
            // left subtreee;
            root.left = insert(root.left, val);
        } else {
            root.right = insert(root.right, val);
        }
        return root;
    }

    public static void inorder(Node root) {
        if (root == null) {
            return;
        }
        inorder(root.left);
        System.out.print(root.data + " ");
        inorder(root.right);
    }

    public static void PrintInRange(Node root, int k1, int k2) {
        if (root.data >= k1 && root.data <= k2) {
            PrintInRange(root.left, k1, k2);
            System.out.print(root.data + " ");
            PrintInRange(root.right, k1, k2);
        } else if (root.data < k1) {
            PrintInRange(root.left, k1, k2);
        } else {
            PrintInRange(root.right, k1, k2);
        }
    }

    public static Node delete(Node root, int val) {
        if (root.data < val) {
            root.right = delete(root.right, val);
        } else if (root.data > val) {
            root.left = delete(root.left, val);
        } else {
            // case 1:// 1--> 1 leaf node
            if (root.left == null && root.right == null) {
                return null;
            }

            // cases 2. single child

            if (root.left == null) {
                return root.right;
            } else if (root.right == null) {
                return root.left;
            }

            Node Is = findInorderSucessor(root.right);
            root.data = Is.data;

            root.right = delete(root.right, root.data);
        }
        return root;

    }

    public static Node findInorderSucessor(Node root) {
        while (root.left != null) {
            root = root.left;

        }
        return root;
    }

    public static void printPath(ArrayList<Integer> path) {
        for (int i = 0; i < path.size(); i++) {
            System.out.print(path.get(i) + "->");
        }
        System.out.println("NULL");

    }

    public static void printRoot2Leaf(Node root, ArrayList<Integer> path) {
        if (root == null) {
            return;
        }
        path.add(root.data);
        if (root.left == null && root.right == null) {
            printPath(path);
        }
        printRoot2Leaf(root.left, path);
        printRoot2Leaf(root.right, path);
        path.remove(path.size() - 1);
        return;

    }

    public static void main(String args[]) {
        int values[] = { 5, 1, 3, 4, 2, 7 };
        Node root = null;
        for (int i = 0; i < values.length; i++) {
            root = insert(root, values[i]);
        }
        System.out.println("Before delete");
        inorder(root);
        delete(root, 5);
        System.out.println("\nAfter delete");
        inorder(root);
        System.out.println();
        // PrintInRange(root, 5, 11);
        printRoot2Leaf(root, new ArrayList<>());

    }

}
